framework/include/utils/CellCenteredMapFunctor.h

```diff
@@ -0,0 +1,145 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseFunctor.h"
+#include "GreenGaussGradient.h"
+#include "MathFVUtils.h"
+#include "libmesh/utility.h"
+#include "libmesh/type_tensor.h"
+#include "libmesh/compare_types.h"
+#include "libmesh/threads.h"
+
+template <typename T, typename T2, typename std::enable_if<ScalarTraits<T>::value, int>::type = 0>
+inline TypeVector<typename CompareTypes<T, T2>::supertype>
+outer_product(const T & a, const TypeVector<T2> & b)
+{
+  TypeVector<typename CompareTypes<T, T2>::supertype> ret;
+  for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+    ret(i) = a * b(i);
+
+  return ret;
+}
+
+template <typename T, typename Map>
+class CellCenteredMapFunctor : public Moose::FunctorImpl<T>, public Map
+{
+public:
+  using typename Moose::FunctorImpl<T>::FaceArg;
+  using typename Moose::FunctorImpl<T>::SingleSidedFaceArg;
+  using typename Moose::FunctorImpl<T>::ElemFromFaceArg;
+  using typename Moose::FunctorImpl<T>::ElemQpArg;
+  using typename Moose::FunctorImpl<T>::ElemSideQpArg;
+  using typename Moose::FunctorImpl<T>::ValueType;
+  using typename Moose::FunctorImpl<T>::GradientType;
+  using typename Moose::FunctorImpl<T>::DotType;
+
+  CellCenteredMapFunctor(const MooseMesh & mesh,
+                         const bool nonorthgonal_correction,
+                         const bool correct_skewness = false,
+                         const bool map_filled = false)
+    : _mesh(mesh),
+      _nonorthgonal_correction(nonorthgonal_correction),
+      _correct_skewness(correct_skewness),
+      _map_filled(map_filled)
+  {
+  }
+
+  bool isExtrapolatedBoundaryFace(const FaceInfo & fi) const override { return !fi.neighborPtr(); }
+
+  using Moose::FunctorImpl<T>::operator();
+  ValueType operator()(const FaceInfo & fi) const { return (*this)(makeFace(fi)); }
+
+  using Moose::FunctorImpl<T>::gradient;
+  GradientType gradient(const FaceInfo & fi) const { return gradient(makeFace(fi)); }
+
+  void mapFilled(const bool map_filled) { _map_filled = map_filled; }
+
+private:
+  FaceArg makeFace(const FaceInfo & fi) const { return Moose::FV::makeCDFace(fi); }
+
+  const MooseMesh & _mesh;
+  const bool _nonorthgonal_correction;
+  const bool _correct_skewness;
+  bool _map_filled;
+  mutable Threads::spin_mutex _map_mutex;
+
+  ValueType evaluate(const libMesh::Elem * const & elem, unsigned int) const override final
+  {
+    if (_map_filled)
+      return libmesh_map_find(*this, elem->id());
+
+    Threads::spin_mutex::scoped_lock lock(_map_mutex);
+    return (*const_cast<CellCenteredMapFunctor *>(this))[elem->id()];
+  }
+
+  ValueType evaluate(const FaceArg & face, unsigned int) const override final
+  {
+    const auto & fi = *std::get<0>(face);
+    const auto elem_value = (*this)(&fi.elem());
+    if (fi.neighborPtr())
+      return fi.gC() * elem_value + (1 - fi.gC()) * (*this)(fi.neighborPtr());
```

Comment:
This code has since been changed and now uses a new `linearInterpolation` overload which can perform skew corrections ... based off information contained in the `FaceArg`. `FaceArg` is now a bonafide `struct` and instead of using `std::get` to obtain pieces of it, you now very transparently query plain-old data like `face.correct_skewness` or `face.fi`

Reply:
Ahhh, super nice!